### PR TITLE
implement ingress support

### PR DESCRIPTION
**Pull request by linki**
very hacky but functional

### CODE REVIEW

**Comment by linki** [comment]
we could split it into two `Producer`s - one for services, one for ingress both implementing the `Producer` interface - and then implement the `Kubernetes` producer by calling them and also implementing the `Producer` interface.

File: producers/kubernetes.go [hunk 1]
```diff
@@ -177,3 +245,44 @@ func (a *kubernetesProducer) convertToEndpoint(svc api.Service) (*pkg.Endpoint,
 
 	return ep, nil
 }
+
+func validateIngress(ing extensions.Ingress) (bool, string) {
```
Comment (ideahitme) [inline producers/kubernetes.go]: why not return `error` ?

File: producers/kubernetes.go [hunk 2]
```diff
@@ -177,3 +245,44 @@ func (a *kubernetesProducer) convertToEndpoint(svc api.Service) (*pkg.Endpoint,
...
+		return false, fmt.Sprintf(
+			"Cannot register ingress '%s/%s'. More than one ingress is not supported",
+			ing.Namespace, ing.Name,
+		)
+	}
+
+	return true, ""
+}
+
+func (a *kubernetesProducer) convertIngressToEndpoint(ing extensions.Ingress) ([]*pkg.Endpoint, error) {
```
Comment (ideahitme) [inline producers/kubernetes.go]: then maybe the other func should be convertServiceToEndpoint

Comment (linki) [reply producers/kubernetes.go]: yes, or split the struct and make them local

File: producers/kubernetes.go (hunk 1 above)
Comment (linki) [reply producers/kubernetes.go]: no idea, this code needs some love

File: producers/kubernetes.go [hunk 3]
```diff
@@ -62,14 +63,15 @@ func NewKubernetes() (*kubernetesProducer, error) {
 }
 
 func (a *kubernetesProducer) Endpoints() ([]*pkg.Endpoint, error) {
+	ret := make([]*pkg.Endpoint, 0)
```
Comment (szuecs) [inline producers/kubernetes.go]: Why this is called ret and not endpoints?

Comment (ideahitme) [reply producers/kubernetes.go]: @szuecs probably because it is to be `ret`urned

Comment (linki) [reply producers/kubernetes.go]: @szuecs yeah, it's silly but we should go over these things in another PR

Comment (szuecs) [reply producers/kubernetes.go]: @linki I have to disagree

**Comment by linki** [comment]
extracted one producer for both services and ingress. then created another producer that aggregates both. LMKWYT @ideahitme

**Comment by ideahitme** [comment]
@linki reviewing :)

**Comment by ideahitme** [comment]
@linki idea of having pluggable ingress/service producer is nice. Since for now are not really sure if we are going to support ingress DNS record generation, having current implementation is okay. But going further, we could have improved the structure by allowing each producer independent of another (better structure, better error handling through goroutines, CLI flags etc. ) 

We can improve the structure going forward, your thoughts ?

**Comment by linki** [comment]
@ideahitme i totally agree.

also, after thinking about it more, i agree with your proposal of being able to configure multiple producers and consumers. it will be backwards compatible (assuming we provide multiple producers with repeated `--producer` flags), we can still have both individual service/ingress producers as well as a combined one, and it will be a nice programming exercise dealing with more and diverse things running concurrently.

**Comment by ideahitme** [comment]
👍

**Pull request merged**